'ps.h' file not found

Hi, do you know where could I get this `ps.h`? I guess it's a header of postscript. It was from a clean docker container `ubuntu:18.04`. 

```bash
docker run -it --rm --name js -v "$PWD:/opt/work" -w "/opt/work" -p 8080:8080 ubuntu:18.04 bash
```
And couple instructions.
```bash
apt update
./scripts/cpp-install-prerequisites.sh
npm ci
npm run install-build-deps
npm run build
```

```bash
[ 97%] Building C object graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/opt/work/src-graphviz/plugin/core/gvrender_core_tk.c.o
[ 98%] Building C object graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/opt/work/src-graphviz/plugin/core/gvrender_core_vml.c.o
/opt/work/src-graphviz/plugin/core/gvrender_core_ps.c:24:10: fatal error: 'ps.h' file not found
#include "ps.h"
         ^~~~~~
1 error generated.
shared:ERROR: '/opt/work/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=1 -D__EMSCRIPTEN_minor__=39 -D__EMSCRIPTEN_tiny__=8 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration -Xclang -nostdsysteminc -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/include/libcxx -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/lib/libcxxabi/include -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/include/compat -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/include -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/include/libc -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/lib/libc/musl/arch/emscripten -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/local/include -Xclang -isystem/root/.emscripten_cache/wasm-obj/include -I/opt/work/cpp/graphviz -I/opt/work/cpp/../src-graphviz/build/plugin/core -I/opt/work/cpp/../src-graphviz/plugin/core -I/opt/work/cpp/../src-graphviz/lib/vpsc -I/opt/work/cpp/../src-graphviz/lib/twopigen -I/opt/work/cpp/../src-graphviz/lib/sparse -I/opt/work/cpp/../src-graphviz/lib/sfdpgen -I/opt/work/cpp/../src-graphviz/lib/pathplan -I/opt/work/cpp/../src-graphviz/lib/common -I/opt/work/cpp/../src-graphviz/lib/gvc -I/opt/work/cpp/../src-graphviz/lib/cdt -I/opt/work/cpp/../src-graphviz/lib/cgraph -I/opt/work/cpp/../src-graphviz/lib/xdot -DNDEBUG -Os -c -DEMSCRIPTEN /opt/work/src-graphviz/plugin/core/gvrender_core_ps.c -fno-inline-functions -Xclang -isystem/opt/work/emsdk/upstream/emscripten/system/include/SDL -c -o CMakeFiles/gvplugin_core.dir/opt/work/src-graphviz/plugin/core/gvrender_core_ps.c.o -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr' failed (1)
graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/build.make:238: recipe for target 'graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/opt/work/src-graphviz/plugin/core/gvrender_core_ps.c.o' failed
make[2]: *** [graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/opt/work/src-graphviz/plugin/core/gvrender_core_ps.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:1333: recipe for target 'graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/all' failed
make[1]: *** [graphviz/plugin/core/CMakeFiles/gvplugin_core.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

`ubuntu:1804` doesn't have `wget` by default. It needs to be installed before `./scripts/cpp-install-graphviz.sh`. 
It could compile entire graphviz now. But I got some undefined errors during generating webassembly.
```bash
[ 98%] Built target gvplugin_core
[ 98%] Generating main_glue.js, main_glue.cpp
Scanning dependencies of target graphvizlib
[ 99%] Building CXX object graphviz/graphvizlib/CMakeFiles/graphvizlib.dir/main.cpp.o
[ 99%] Building CXX object graphviz/graphvizlib/CMakeFiles/graphvizlib.dir/fs.cpp.o
[100%] Linking CXX executable graphvizlib.js
cache:INFO: generating system library: libcompiler_rt.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libcompiler_rt.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc-wasm.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libc-wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc++-noexcept.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libc++-noexcept.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc++abi-noexcept.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libc++abi-noexcept.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libdlmalloc.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libdlmalloc.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libpthread_stub.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libpthread_stub.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libc_rt_wasm.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libc_rt_wasm.a" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: libsockets.a... (this will be cached in "/root/.emscripten_cache/wasm-obj/libsockets.a" for subsequent builds)
cache:INFO:  - ok
wasm-ld: error: ../plugin/core/libgvplugin_core.a(gvrender_core_dot.c.o): undefined symbol: AgIoDisc
wasm-ld: error: ../plugin/core/libgvplugin_core.a(gvrender_core_dot.c.o): undefined symbol: AgIoDisc
wasm-ld: error: ../plugin/neato_layout/libgvplugin_neato_layout.a(gvlayout_neato_layout.c.o): undefined symbol: Nop
wasm-ld: error: ../plugin/neato_layout/libgvplugin_neato_layout.a(gvlayout_neato_layout.c.o): undefined symbol: Nop
wasm-ld: error: ../plugin/neato_layout/libgvplugin_neato_layout.a(gvlayout_neato_layout.c.o): undefined symbol: Nop
wasm-ld: error: ../plugin/neato_layout/libgvplugin_neato_layout.a(gvlayout_neato_layout.c.o): undefined symbol: Nop
wasm-ld: error: ../lib/circogen/libcircogen.a(blocktree.c.o): undefined symbol: Verbose
wasm-ld: error: ../lib/circogen/libcircogen.a(circularinit.c.o): undefined symbol: Ndim
wasm-ld: error: ../lib/circogen/libcircogen.a(circularinit.c.o): undefined symbol: E_weight
wasm-ld: error: ../lib/circogen/libcircogen.a(circularinit.c.o): undefined symbol: Agstrictundirected
wasm-ld: error: ../lib/circogen/libcircogen.a(circularinit.c.o): undefined symbol: Ndim
wasm-ld: error: ../lib/dotgen/libdotgen.a(class2.c.o): undefined symbol: Concentrate
wasm-ld: error: ../lib/dotgen/libdotgen.a(class2.c.o): undefined symbol: Concentrate
wasm-ld: error: ../lib/dotgen/libdotgen.a(class1.c.o): undefined symbol: E_constr
wasm-ld: error: ../lib/circogen/libcircogen.a(blocktree.c.o): undefined symbol: Verbose
wasm-ld: error: ../lib/circogen/libcircogen.a(circularinit.c.o): undefined symbol: E_weight
wasm-ld: error: ../lib/dotgen/libdotgen.a(dotinit.c.o): undefined symbol: N_group
wasm-ld: error: ../lib/dotgen/libdotgen.a(dotinit.c.o): undefined symbol: N_group
wasm-ld: error: ../lib/dotgen/libdotgen.a(dotinit.c.o): undefined symbol: E_showboxes
wasm-ld: error: ../lib/dotgen/libdotgen.a(dotinit.c.o): undefined symbol: E_minlen
wasm-ld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
shared:ERROR: '/opt/work/emsdk/upstream/bin/wasm-ld -o /opt/work/emscripten_temp_aajUPp/graphvizlib.wasm --lto-O0 CMakeFiles/graphvizlib.dir/main.cpp.o CMakeFiles/graphvizlib.dir/fs.cpp.o -L/opt/work/emsdk/upstream/emscripten/system/local/lib ../lib/gvc/libgvc.a -L/opt/work/emsdk/upstream/emscripten/system/lib ../plugin/core/libgvplugin_core.a -L/root/.emscripten_cache/wasm-obj ../plugin/dot_layout/libgvplugin_dot_layout.a ../plugin/neato_layout/libgvplugin_neato_layout.a ../lib/circogen/libcircogen.a ../lib/dotgen/libdotgen.a ../lib/osage/libosage.a ../lib/patchwork/libpatchwork.a ../lib/twopigen/libtwopigen.a ../lib/sfdpgen/libsfdpgen.a ../lib/gvc/libgvc.a ../lib/common/libcommon.a ../lib/cgraph/libcgraph.a ../lib/pathplan/libpathplan.a ../lib/sparse/libsparse.a ../lib/xdot/libxdot.a ../lib/neatogen/libneatogen.a ../lib/pack/libpack.a ../lib/ortho/libortho.a ../lib/fdpgen/libfdpgen.a ../lib/label/liblabel.a ../lib/gvc/libgvc.a ../lib/common/libcommon.a ../lib/cgraph/libcgraph.a ../lib/pathplan/libpathplan.a ../lib/sparse/libsparse.a ../lib/xdot/libxdot.a ../lib/neatogen/libneatogen.a ../lib/pack/libpack.a ../lib/ortho/libortho.a ../lib/fdpgen/libfdpgen.a ../lib/label/liblabel.a ../lib/expat/libexpat.a ../lib/vpsc/libvpsc.a ../lib/cdt/libcdt.a /root/.emscripten_cache/wasm-obj/libc.a /root/.emscripten_cache/wasm-obj/libcompiler_rt.a /root/.emscripten_cache/wasm-obj/libc-wasm.a /root/.emscripten_cache/wasm-obj/libc++-noexcept.a /root/.emscripten_cache/wasm-obj/libc++abi-noexcept.a /root/.emscripten_cache/wasm-obj/libdlmalloc.a /root/.emscripten_cache/wasm-obj/libpthread_stub.a /root/.emscripten_cache/wasm-obj/libc_rt_wasm.a /root/.emscripten_cache/wasm-obj/libsockets.a --allow-undefined --import-memory --import-table -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --strip-debug --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location --export emscripten_builtin_memalign --export memalign --export emscripten_builtin_free --export realloc --export _get_environ --export _ZSt18uncaught_exceptionv -z stack-size=5242880 --initial-memory=16777216 --no-entry --global-base=1024' failed (1)
graphviz/graphvizlib/CMakeFiles/graphvizlib.dir/build.make:167: recipe for target 'graphviz/graphvizlib/graphvizlib.js' failed
make[2]: *** [graphviz/graphvizlib/graphvizlib.js] Error 1
CMakeFiles/Makefile2:1520: recipe for target 'graphviz/graphvizlib/CMakeFiles/graphvizlib.dir/all' failed
make[1]: *** [graphviz/graphvizlib/CMakeFiles/graphvizlib.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/.dockerignore
+++ b/.dockerignore
@@ -1,0 +1,15 @@
+.vscode/ipch
+build/
+dist/
+emsdk/
+src-expat/
+src-graphviz/
+lib-*/
+node_modules/
+types/
+*.tsbuildinfo
+.nyc_output/
+coverage/
+.vscode/c_cpp_properties.json
+.vscode/settings.json
+test/ubuntu-dev.dockerfile

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,9 +11,10 @@ env:
     - EM_USE_GLOBAL_CACHE=1
 install:
   - npm ci
-  - npm run install-prerequisites
+  - sudo apt-get install -y git cmake wget
+  - sudo apt-get install -y gcc-multilib g++-multilib pkg-config autoconf bison libtool flex zlib1g-dev
+  - sudo apt-get install -y python2.7 python-pip
   - npm run install-build-deps
 before_script:
   - npm run lint
-  - npm run dev-start &
   - npm run build

--- a/README.md
+++ b/README.md
@@ -202,13 +202,19 @@ Convenience function that performs **patchwork** layout, is equivalent to `layou
 Convenience function that performs **twopi** layout, is equivalent to `layout(dotSource, outputFormat, "twopi");`.
 
 ## Building @hpcc-js/wasm
-_Building is supported on both Linux (tested with Ubuntu 18.04) and Windows with WSL enabled (Ubuntu-18.04).  Building in other environments (docker) should work, but may be missing certain prerequisites, should you encounter any please open an issue with the details so that the pre-requisites file can be updated._
+_Building is supported on both Linux (tested with Ubuntu 18.04) and Windows with WSL enabled (Ubuntu-18.04).  Building in other environments should work, but may be missing certain prerequisites._
 
-There are several required OS dependencies which can be installed with:
+These are then known required OS dependencies:
 ```
-sudo ./scripts/cpp-install-prerequisites.sh
+sudo apt-get install -y curl
+sudo curl --silent --location https://deb.nodesource.com/setup_12.x | sudo bash -
+sudo apt-get install -y nodejs
+sudo apt-get install -y build-essential
+
+sudo apt-get install -y git cmake wget
+sudo apt-get install -y gcc-multilib g++-multilib pkg-config autoconf bison libtool flex zlib1g-dev 
+sudo apt-get install -y python2.7 python-pip
 ```
-(See [./scripts/cpp-install-prerequisites.sh](./scripts/cpp-install-prerequisites.sh) for details)
 
 ### Build steps:
 ```
@@ -231,4 +237,3 @@ _It is worth noting that `npm run clean` will only clean any artifacts associate
 ```
 npm run clean-build-deps
 ```
-

--- a/package.json
+++ b/package.json
@@ -12,9 +12,6 @@
     "src/*.css"
   ],
   "scripts": {
-    "install-prerequisites": "run-script-os",
-    "install-prerequisites:linux": "echo nop",
-    "install-prerequisites:win32": "wsl -u root -e ./scripts/cpp-install-prerequisites.sh",
     "install-emsdk": "run-script-os",
     "install-emsdk:linux": "./scripts/cpp-install-emsdk.sh",
     "install-emsdk:win32": "wsl -e ./scripts/cpp-install-emsdk.sh",
@@ -40,13 +37,15 @@
     "bundle-watch": "npm run bundle -- -w",
     "minimize": "terser dist/index.js -c -m --source-map -o dist/index.min.js",
     "build": "run-s compile-cpp compile-es6 bundle minimize",
+    "build-ubuntu-dev": "docker build --rm -f \"./test/ubuntu-dev.dockerfile\" -t ubuntu-dev:latest \".\"",
     "publish-version": "standard-version",
     "publish-push": "git push --follow-tags upstream",
     "prepublishOnly": "run-s clean build publish-version publish-push",
     "watch": "run-p compile-es6-watch bundle-watch",
     "dev-start": "ws",
     "lint": "tslint --project . src/**/*.ts",
-    "test": "mocha-chrome http://localhost:8000/test.html"
+    "mocha-test": "mocha-chrome http://localhost:8000/test.html",
+    "test": "run-p -r dev-start mocha-test"
   },
   "dependencies": {},
   "devDependencies": {

--- a/scripts/cpp-install-prerequisites.sh
+++ b/scripts/cpp-install-prerequisites.sh
@@ -1,5 +1,0 @@
-#!/bin/bash
-
-apt install -y nodejs npm git-core cmake wget
-apt install -y gcc-multilib g++-multilib pkg-config autoconf bison libtool flex zlib1g-dev 
-apt install -y python2.7 python-pip

--- a/test/ubuntu-dev.dockerfile
+++ b/test/ubuntu-dev.dockerfile
@@ -1,0 +1,29 @@
+FROM ubuntu:18.04
+
+RUN apt-get update
+RUN apt-get upgrade -y
+
+# Install pre-requisites (keep synced with README.md)
+## NodeJS
+RUN apt-get install -y curl
+RUN curl --silent --location https://deb.nodesource.com/setup_12.x | bash -
+RUN apt-get install -y nodejs
+RUN apt-get install -y build-essential
+
+## Other
+RUN apt-get install -y git cmake wget
+RUN apt-get install -y gcc-multilib g++-multilib pkg-config autoconf bison libtool flex zlib1g-dev 
+RUN apt-get install -y python2.7 python-pip
+
+# Set the working directory
+WORKDIR /usr/src/app
+
+COPY ./scripts/* ./scripts/
+RUN ./scripts/cpp-install-emsdk.sh
+RUN ./scripts/cpp-install-graphviz.sh
+RUN ./scripts/cpp-install-expat.sh
+
+COPY . .
+
+RUN npm ci
+RUN npm run build

